File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.java
class Solution {
    int maxTime = Integer.MIN_VALUE;
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>(n);
        for(int i=0; i<n; i++) adj.add(new ArrayList<Integer>());
        for(int i=0; i<n; i++){
            if(manager[i] != -1) adj.get(manager[i]).add(i);
        }
        dfs(adj, informTime, headID, 0);
        return maxTime;
    }
    void dfs(ArrayList<ArrayList<Integer>> adj, int[] informTime, int curr, int time){
        maxTime = Math.max(maxTime, time);
        for(int x : adj.get(curr)) dfs(adj, informTime, x, time + informTime[curr]);
    }
}

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.py
class Solution:
    maxTime = float('-inf')
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        adj  = [[] for _ in range(n)]
        for i in range(n):
            if manager[i] != -1:
                adj[manager[i]].append(i)
        self.dfs(adj, informTime, headID, 0)
        return self.maxTime
    def dfs(self, adj, informTime, curr, time):
        self.maxTime = max(self.maxTime, time)
        for x in adj[curr]:
                self.dfs(adj, informTime, x, time+informTime[curr])